File: core/src/main/java/heroes/journey/utils/worldgen/FeatureType.java
package heroes.journey.utils.worldgen;

import static heroes.journey.registries.Registries.FeatureTypeManager;

import heroes.journey.GameState;
import heroes.journey.entities.Position;
import heroes.journey.registries.Registrable;
import heroes.journey.tilemap.Feature;
import lombok.Getter;

@Getter
public abstract class FeatureType extends Registrable {

    public FeatureType(String id, String name) {
        super(id, name);
        FeatureTypeManager.register(this);
    }

    public abstract Feature generateFeature(GameState gs, Position pos, Feature... connections);

}

File: core/src/main/java/heroes/journey/utils/worldgen/FeatureType.kt
package heroes.journey.utils.worldgen

import heroes.journey.GameState
import heroes.journey.entities.Position
import heroes.journey.registries.Registrable
import heroes.journey.registries.Registries.FeatureTypeManager
import heroes.journey.tilemap.Feature
import lombok.Getter

@Getter
abstract class FeatureType(id: String, name: String?) : Registrable(id, name) {

    override fun register(): FeatureType {
        return FeatureTypeManager.register(this)
    }

    abstract fun generateFeature(gs: GameState, pos: Position, vararg connections: Feature): Feature
}
